### PR TITLE
Add flatten option to the NLocal family

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

This commit adds a new flag `flatten` to the circuit library elements in the NLocal family. This flag is used to avoid the nested wrapping of the output that is done by default in the output of these gates. This is done primarily in the interest of performance. The runtime performance of bind_parameters() is vastly improved with a flattened output compared to the nested object. For example, running:

```
qc = EfficientSU2(100, entanglement="linear", reps=100)
qc.measure_all()
qc.bind_parameters({x: math.pi / 2 for x in qc.parameters})
```

the runtime of `bind_parameters()` went from ~390 seconds with the default wrapped output to ~0.5 seconds with the flattened output. I think given these results longer term we might want to flip the default behavior to only wrap on user request (likely for visualization as that's the only use case I can think of for the wrapped output). The default value is set to `None` in this PR to facilitate this change. In 0.26/0.45 we can emit a warning if `flatten` is `None` to warn that the default will change in a future release.

### Details and comments

### CODE REVIEW

**Comment by woodsp-ibm** [comment]
A couple of thoughts: 

Would it make sense to expose flatten as a property too so that you can change how it comes out for visualization without having to construct a whole new identical circuit? In notebooks we show an ansatz (or a combination of feature map and anstaz in ML) - sometimes decomposed at present  sometimes not depending on what we want to show. Changing flatten would invalidate it so it needs to get built again.

I wonder whether this could/should be done for data_preparation (feature maps) too.

QAOAnsatz is not directly an NLocal cct but goes through EvolvedOperatorAnsatz (which UCC in Nature is based on). I saw that was not changed but further see the tests with qaoa failed around passing flatten to the parent.

**Comment by mtreinish** [comment]
> A couple of thoughts:
> 
> Would it make sense to expose flatten as a property too so that you can change how it comes out for visualization without having to construct a whole new identical circuit? In notebooks we show an ansatz (or a combination of feature map and anstaz in ML) - sometimes decomposed at present sometimes not depending on what we want to show. Changing flatten would invalidate it so it needs to get built again.

I worry about changing the contents of the circuit dynamically like this if feels kind of error prone (granted I feel that way about blueprint circuits in general though). Maybe it's as easy as setting `_is_built` to false in the setter. I'll test it locally

> 
> I wonder whether this could/should be done for data_preparation (feature maps) too.

Yeah I started small here, but I think we'll want to do this to all the circuit library classes that have wrapped output. The performance difference is so drastically improved. I was tempted to just revert #6634 at first. But figured this approach was a bit more sustainable than a breaking api change revert of a breaking change from 2 yrs ago.

> 
> QAOAnsatz is not directly an NLocal cct but goes through EvolvedOperatorAnsatz (which UCC in Nature is based on). I saw that was not changed but further see the tests with qaoa failed around passing flatten to the parent.

Yeah, I reverted this when I saw. I was just confused because it's in the nlocal directory and was kind of on autopilot when I was making the changes.

File: qiskit/circuit/library/n_local/n_local.py [hunk 1]
```diff
@@ -943,12 +955,13 @@ def _build(self) -> None:
...
-        except QiskitError:
-            block = circuit.to_instruction()
+        if not self._flatten:
+            try:
+                block = circuit.to_gate()
+            except QiskitError:
+                block = circuit.to_instruction()
 
-        self.append(block, self.qubits)
+            self.append(block, self.qubits)
```
[the remark this answers is not in the document]
Comment (jakelishman) [reply qiskit/circuit/library/n_local/n_local.py:964]: Oh no sorry, my fault - I'd not realised that this was part of the same code block that switched the definition of `circuit` to `self` above.

**Comment by Cryoris** [comment]
This is a good addition 👍🏻 But the current implementation doesn't quite do what I expected from the documentation. For example, the `EfficientSU2` nicely gets decomposed in standard gates:
```python
>>> EfficientSU2(3, flatten=True, reps=1).draw()
     ┌──────────┐┌──────────┐                  ┌──────────┐ ┌──────────┐
q_0: ┤ Ry(θ[0]) ├┤ Rz(θ[3]) ├──────────■───────┤ Ry(θ[6]) ├─┤ Rz(θ[9]) ├
     ├──────────┤├──────────┤        ┌─┴─┐     ├──────────┤┌┴──────────┤
q_1: ┤ Ry(θ[1]) ├┤ Rz(θ[4]) ├──■─────┤ X ├─────┤ Ry(θ[7]) ├┤ Rz(θ[10]) ├
     ├──────────┤├──────────┤┌─┴─┐┌──┴───┴───┐┌┴──────────┤└───────────┘
q_2: ┤ Ry(θ[2]) ├┤ Rz(θ[5]) ├┤ X ├┤ Ry(θ[8]) ├┤ Rz(θ[11]) ├─────────────
     └──────────┘└──────────┘└───┘└──────────┘└───────────┘
```
However, more complex circuits, such as the `EvolvedOperatorAnsatz` don't seem "flattened" but just less convolved:
```python
>>> EvolvedOperatorAnsatz(op, flatten=True).draw()
     ┌─────────────────────────────┐
q_0: ┤0                            ├
     │                             │
q_1: ┤1 exp(-it (ZZZ + XXX))(t[0]) ├
     │                             │
q_2: ┤2                            ├
     └─────────────────────────────┘
```

For an argument called "flatten" I would rather expect something that decomposes to standard gates, no matter the nesting:
```python
# on this circuit, h rz cx are enough, but this should include all standard gates
>>> transpile(EvolvedOperatorAnsatz(op), basis_gates=["h", "rz", "cx"]).draw() 
          ┌───┐┌──────────────┐┌───┐┌───┐          ┌───┐┌──────────────┐┌───┐┌───┐
q_0: ─────┤ X ├┤ Rz(2.0*t[0]) ├┤ X ├┤ H ├──────────┤ X ├┤ Rz(2.0*t[0]) ├┤ X ├┤ H ├─────
     ┌───┐└─┬─┘└──────────────┘└─┬─┘├───┤┌───┐┌───┐└─┬─┘└──────────────┘└─┬─┘├───┤┌───┐
q_1: ┤ X ├──■────────────────────■──┤ X ├┤ H ├┤ X ├──■────────────────────■──┤ X ├┤ H ├
     └─┬─┘                          └─┬─┘├───┤└─┬─┘                          └─┬─┘├───┤
q_2: ──■──────────────────────────────■──┤ H ├──■──────────────────────────────■──┤ H ├
                                         └───┘                                    └───┘
```

That's analog to what `numpy.flatten` does -- no matter the dimension of your tensor, a flattened array is a vector.

So if this argument is supposed to "flatten", would it be better to decompose to the standard gates? Otherwise we could also rename it to `wrap` (default=True), to make it clearer?

**Comment by mtreinish** [comment]
> So if this argument is supposed to "flatten", would it be better to decompose to the standard gates? Otherwise we could also rename it to `wrap` (default=True), to make it clearer?

I fixed the flatten behavior for `EvolvedOperatorAnsatz` here: https://github.com/Qiskit/qiskit-terra/pull/10269/commits/b56e960ee45d4bd562f8e457215e54ae2adcfc09

The tricky bit for that class was that it was built using other circuit library elements which were proper gates, so from the perspective of the flatten option it was flattened to just gate objects (instead of wrapped circuits containing other circuit library elements).

File: qiskit/circuit/library/evolved_operator_ansatz.py [hunk 2]
```diff
@@ -40,6 +41,7 @@ def __init__(
         name: str = "EvolvedOps",
         parameter_prefix: str | Sequence[str] = "t",
         initial_state: QuantumCircuit | None = None,
+        flatten: Optional[bool] = None,
```
Comment (Cryoris) [inline qiskit/circuit/library/evolved_operator_ansatz.py:44]: For consistency with the rest of the signature (also the `Optional` import would need to be removed 🙂)
```suggestion
        flatten: bool | None = None,
```

File: qiskit/circuit/library/n_local/n_local.py [hunk 3]
```diff
@@ -90,6 +90,7 @@ def __init__(
         skip_unentangled_qubits: bool = False,
         initial_state: QuantumCircuit | None = None,
         name: str | None = "nlocal",
+        flatten: Optional[bool] = None,
```
Comment (Cryoris) [inline qiskit/circuit/library/n_local/n_local.py:93]:
```suggestion
        flatten: bool | None = None,
```

File: qiskit/circuit/library/n_local/qaoa_ansatz.py [hunk 4]
```diff
@@ -39,6 +41,7 @@ def __init__(
         initial_state: QuantumCircuit | None = None,
         mixer_operator=None,
         name: str = "QAOA",
+        flatten: Optional[bool] = None,
```
Comment (Cryoris) [inline qiskit/circuit/library/n_local/qaoa_ansatz.py:44]:
```suggestion
        flatten: bool | None = None,
```

**Review by Cryoris — APPROVED** [review]
LGTM, just tiny comments on consistency on the signature that would be nice to get in, but not required.

**Comment by mtreinish** [comment]
In the interest of time I've enqueued it to merge I'll open an issue based on your comment and track that as a fix post-rc1 (since it's just a doc change)

**Pull request merged**